apis/iam/v1beta1/diff.go

```diff
@@ -0,0 +1,157 @@
+	}
+	if desired.ResourceReference.Namespace != actual.ResourceReference.Namespace {
+		diff.AddField("spec.resourceRef.namespace", desired.ResourceReference.Namespace, actual.ResourceReference.Namespace)
+	}
+	if desired.ResourceReference.Name != actual.ResourceReference.Name {
+		diff.AddField("spec.resourceRef.name", desired.ResourceReference.Name, actual.ResourceReference.Name)
+	}
+	if desired.ResourceReference.External != actual.ResourceReference.External {
+		diff.AddField("spec.resourceRef.external", desired.ResourceReference.External, actual.ResourceReference.External)
+	}
+
+	// Compare Bindings
+	if bindingsDiffer(desired.Bindings, actual.Bindings, diff) {
+	}
+
+	// Compare AuditConfigs
+	if auditConfigsDiffer(desired.AuditConfigs, actual.AuditConfigs, diff) {
+	}
+
+	return len(diff.Fields) > 0
+}
+
+// bindingsDiffer compares two slices of IAMPolicyBinding.
+// It treats the slices as maps keyed by Role, and members are treated as unordered sets.
+func bindingsDiffer(desired, actual []IAMPolicyBinding, diff *structuredreporting.Diff) bool {
+	changed := false
+	desiredMap := make(map[string]IAMPolicyBinding)
+	for _, b := range desired {
+		sort.Slice(b.Members, func(i, j int) bool {
+			return b.Members[i] < b.Members[j]
+		})
+		// Canonicalize members for stable comparison.
+		desiredMap[b.Role] = b
+	}
+
+	actualMap := make(map[string]IAMPolicyBinding)
+	for _, b := range actual {
+		sort.Slice(b.Members, func(i, j int) bool {
+			return b.Members[i] < b.Members[j]
+		})
+		// Canonicalize members for stable comparison.
+		actualMap[b.Role] = b
+	}
+
+	// Check for added or modified bindings.
+	for role, desiredBinding := range desiredMap {
+		actualBinding, ok := actualMap[role]
+		if !ok {
+			diff.AddField(fmt.Sprintf("spec.bindings[role=%v]", role), "present in desired spec", "absent in actual spec")
+			changed = true
+			continue
+		}
+		if !reflect.DeepEqual(desiredBinding, actualBinding) {
+			diff.AddField(fmt.Sprintf("spec.bindings[role=%v]", role), desiredBinding, actualBinding)
+			changed = true
+		}
+	}
+
+	// Check for removed bindings.
+	for role := range actualMap {
+		if _, ok := desiredMap[role]; !ok {
+			diff.AddField(fmt.Sprintf("spec.bindings[role=%v]", role), "absent in desired spec", "present in actual spec")
+			changed = true
+		}
+	}
+	return changed
+}
+
+// auditConfigsDiffer compares two slices of IAMPolicyAuditConfig.
+// It treats the slices as maps keyed by Service and canonicalizes inner slices for stable comparison.
+func auditConfigsDiffer(desired, actual []IAMPolicyAuditConfig, diff *structuredreporting.Diff) bool {
+	changed := false
+	// Helper function to canonicalize the inner slices for a stable comparison.
+	canonicalize := func(configs []IAMPolicyAuditConfig) {
+		// Create a deep copy to avoid modifying the original struct.
+		copiedConfigs := make([]IAMPolicyAuditConfig, len(configs))
```

Comment:
The comment does not match the code. Not understanding the point of making an array and (shallowish) copying into it. Seems like we can just order the input at that point.

Reply:
https://go.dev/play/p/N5slMS4wfok?v=goprev

Reply:
Looking around the callsite for the diff for this function, I think we should keep the idea of deep copying the structure and sorting it, even if we discard it. I wouldn't want the sorting to cause diffs in status reporting or something else down the road.

I would be game to sort these as part of the reconciler loop (if we don't do this already).

wdyt? updated the code.